File: minimumSizeSubarraySum.py
class Solution(object):
    def minSubArrayLen(self, target, nums):
        """
        :type target: int
        :type nums: List[int]
        :rtype: int
        """
        res = len(nums) + 1
        start, cur = -1, 0
        for i, n in enumerate(nums):
            if n >= target:
                return 1
            elif start == -1:
                cur = n
                start = i
            else:
                cur += n
                while cur - nums[start] >= target:
                    cur -= nums[start]
                    start += 1
                if cur >= target:
                    res = min(res,i - start + 1)
        if res == len(nums) + 1 :
            return 0
        return res

File: minimumSizeSubarraySum.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int N = nums.size();
        int si = 0;
        int res = N + 1;
        int cur = 0;
        for (int i = 0; i < N; i++) {
            cur += nums[i];
            if (i - si + 1 >= res) {
                cur -= nums[si++];
            }
            if (cur >= target) {
                while(cur - nums[si] >= target) {
                    cur -= nums[si++];
                }
                res = i - si + 1;
            }
        }
        return res > N ? 0 : res;
    }
};
